ecc_enclave/enclave/shim.h

```diff
@@ -6,27 +6,167 @@
+// void get_tx_id(char* tx_id,
+//     uint32_t max_tx_id_len,
+//     shim_ctx_ptr_t ctx);
+
+// - getTxTimestamp
+// // TODO (possible extensions): enclave has no access to trusted time.  Time
+// //   from client is apriori not trusted either. However, at least client has
+// //   to commit and in some cases might be trusted
+// //   - do endorsers do any cross-check of this value? (Probably makes sense
+// //     only in supporting it if there is some plausibility test the endorsing
+// //     peers agree)
+// //   - Is it commonly used in fabric?
+// #include <time.h>
+// void get_tx_timestamp(struct timespec* ts,
+//     shim_ctx_ptr_t ctx);
+//
+// - getBinding
+// // TODO (possible extensions): from description it seems this is used for replay protection,
+// //   though, from
+// //     https://fabric-shim.github.io/release-1.4/fabric-shim.ChaincodeStub.html#getBinding
+// //   it seems relevant only for some delegation/third-party signature verification
+// //   - Is it commonly used in fabric? If not then we should ignore it
+// //   - Is this something we can easily support (insecurely short-term / securely long-term)?
+//
+// // TODO: other tx-related apis which exist but probably doesn't make sense to support
+// // - getTransient: if we encrypt everything, then everything is essentially Transient?
+// // - getSignedProposal: should be easy to support but probably not worth?
+
+// - creator
+// return the distinguished name of the creator as well as the msp_id of the corresponding
+// organization.
+void get_creator_name(char* msp_id,  // MSP id of organization to which transaction creator belongs
+    uint32_t max_msp_id_len,         // size of allocated buffer for msp_id
+    char* dn,                        // distinguished name of transaction creator
+    uint32_t max_dn_len,             // size of allocated buffer for dn
+    shim_ctx_ptr_t ctx);
+// Note: The name might be truncated (but guaranteed to be null-terminated)
+// if the provided buffer is too small.
+//
+// TODO (eventually): The go shim GoCreator returns protobuf serialized identity which (usally)
+// is the pair of msp_id and a (PEM-encoded) certificate. We might eventually add a function
+// also to expose the certificate itself.  However, for most current use-cases the DN should
+// be sufficient and makes CC-programming easier.
 
-int unmarshal_args(std::vector<std::string>& argss, const char* json_string);
-int unmarshal_values(
-    std::map<std::string, std::string>& values, const char* json_bytes, uint32_t json_len);
+// Chaincode to Chaincode
+//---------------------------
+// invokeChaincode
+// TODO (possible extensions): Currently not supported (but eventually should)
 
-// read/writeset
-void register_rwset(void* ctx, read_set_t* readset, write_set_t* writeset);
-void free_rwset(void* ctx);
-read_set_t* get_read_set(context_t* context, void* ctx);
-write_set_t* get_write_set(context_t* context, void* ctx);
+// logging
```

Comment:
Seems this is aligned with Fabric 1.4.x. This is good.

In Fabric 2.0, however, chaincode logging is the responsibility by the chaincode developer.
https://hyperledger-fabric.readthedocs.io/en/master/logging-control.html

Since this is not easily possible with FPC I still think its a good idea to provide this logging to the user. I believe we can only reduce the number of different log levels we want to support. Maybe just ERROR | WARNING | INFO | DEBUG ?

Reply:
thanks for pointing that out, will consider this when going last stretch with shim revamp and implement logging.  We definitely still should have logging-level edl functions as eventually with our own untrusted stub/shim we still want to be able to route (trusted) shim log-messages to peer.  Given that we do not really have a full libc available, i think we also still would have to provide some logging functionality to user, but then on untrusted side might just throw it directly to stdout (rather than as right now to logger which then passes it to stdout ..)